### PR TITLE
Updated the spec for packaging Foreman on top of a Ruby 1.9.3 SCL

### PR DESCRIPTION
**Pull request by skottler**

### CODE REVIEW

File: foreman.spec [hunk 1]
```diff
@@ -1,23 +1,42 @@
...
+    %global nodoc 1
+%else
+    %global scl_ruby /usr/bin/ruby
+    %global scl_rake /usr/bin/rake
+%endif
 
 Name:   foreman
-Version: 1.1stable
-Release: 3%{dist}
+Version:1.19
```
Comment (domcleal) [inline foreman.spec]: Wrong version number?

File: foreman.spec [hunk 2]
```diff
@@ -1,23 +1,42 @@
...
+Release:1
 Summary:Systems Management web application
 
 Group:  Applications/System
 License:GPLv3+
 URL: http://theforeman.org
-BuildRoot:  %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
-Source0: %{name}-%{version}.tar.gz
+Source0: http://github.com/ohadlevy/%{name}/tarball/%{name}-%{version}.tar.gz
```
Comment (domcleal) [inline foreman.spec]: Wrong URL?

File: foreman.spec [hunk 3]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+BuildRequires: %{?scl_prefix}rubygem(uglifier) >= 1.0.3
+BuildRequires: %{?scl_prefix}rubygem(uuidtools)
+BuildRequires: %{?scl_prefix}rubygem(will_paginate) >= 3.0.2
+BuildRequires: %{?scl_prefix}rubygem(rails)
+BuildRequires: %{?scl_prefix}rubygem(quiet_assets)
+BuildRequires: %{?scl_prefix}rubygem(spice-html5-rails)
+BuildRequires: %{?scl_prefix}rubygem(flot-rails)
+BuildRequires: %{?scl_prefix}facter
+BuildRequires: %{?scl_prefix}puppet >= 0.24.4
+BuildRequires: puppet
```
Comment (domcleal) [inline foreman.spec:109]: Why SCL and non-SCL puppet here?

File: foreman.spec [hunk 4]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(safemode) >= 1.2.0
+Requires: %{?scl_prefix}rubygem(uuidtools)
+Requires: %{?scl_prefix}rubygem(oauth)
+Requires: %{?scl_prefix}rubygem(rabl) >= 0.7.5
+Requires: %{?scl_prefix}rubygem(rake) >= 0.8.3
+Requires: %{?scl_prefix}rubygem(ruby_parser) >= 3.0.0
+Requires: %{?scl_prefix}rubygem(audited-activerecord) >= 3.0.0
+Requires: %{?scl_prefix}rubygem(apipie-rails) >= 0.0.16
+Requires: %{?scl_prefix}rubygem(bundler_ext)
+Requires: %{?scl_prefix}rubygem(thin)
```
Comment (domcleal) [inline foreman.spec]: Is this necessary?

File: foreman.spec [hunk 5]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
-Requires: rubygem(apipie-rails) >= 0.0.13
-Requires: rubygem(oauth)
-Provides: %{name}-%{version}-%{release}
-#Packager:   Ohad Levy <ohadlevy@gmail.com>
+Requires: %{?scl_prefix}rubygem(json)
+Requires: %{?scl_prefix}rubygem(rails) >= 3.2.8
+Requires: %{?scl_prefix}rubygem(rails) < 3.3.0
+Requires: %{?scl_prefix}rubygem(jquery-rails)
+Requires: %{?scl_prefix}rubygem(rest-client)
+Requires: %{?scl_prefix}rubygem(will_paginate) >= 3.0.2
```
Comment (domcleal) [inline foreman.spec]: This is ~> in develop.

File: foreman.spec [hunk 6]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(ancestry) >= 1.3.0
+Requires: %{?scl_prefix}rubygem(ancestry) < 1.4.0
+Requires: %{?scl_prefix}rubygem(scoped_search) >= 2.5.0
+Requires: %{?scl_prefix}rubygem(net-ldap)
+Requires: %{?scl_prefix}rubygem(safemode) >= 1.2.0
```
Comment (domcleal) [inline foreman.spec:59]: This is ~> in develop.

File: foreman.spec [hunk 7]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(uuidtools)
+Requires: %{?scl_prefix}rubygem(oauth)
+Requires: %{?scl_prefix}rubygem(rabl) >= 0.7.5
+Requires: %{?scl_prefix}rubygem(rake) >= 0.8.3
+Requires: %{?scl_prefix}rubygem(ruby_parser) >= 3.0.0
```
Comment (domcleal) [inline foreman.spec:65]: This is ~> in develop.

File: foreman.spec [hunk 8]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(audited-activerecord) >= 3.0.0
```
Comment (domcleal) [inline foreman.spec]: This is currently locked to an RC version due to bug 2393, but I guess leave this as-is for now, as we should fix that.

File: foreman.spec [hunk 9]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(apipie-rails) >= 0.0.16
```
Comment (domcleal) [inline foreman.spec]: This is locked in develop, as it's broken a couple of times in minor releases..

File: foreman.spec [hunk 10]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(bundler_ext)
+Requires: %{?scl_prefix}rubygem(thin)
+Requires: %{?scl_prefix}rubygem(fast_gettext)
```
Comment (domcleal) [inline foreman.spec]: Missing 0.4.8 minimum.

File: foreman.spec [hunk 11]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(gettext_i18n_rails)
+Requires: %{?scl_prefix}rubygem(i18n_data)
```
Comment (domcleal) [inline foreman.spec]: Missing 0.2.6 minimum.

File: foreman.spec [hunk 12]
```diff
@@ -25,35 +44,71 @@ Requires(post): chkconfig
...
+Requires: %{?scl_prefix}rubygem(therubyracer)
```
Comment (domcleal) [inline foreman.spec]: Locked to 0.11.3.

File: foreman.spec [hunk 13]
```diff
@@ -72,16 +127,16 @@ Foreman repository contains open source and other distributable software for
 Fedora. This package contains the repository configuration for Yum.
 
 %files release
-%defattr(-,root,root,-)
-%config(noreplace) %{_sysconfdir}/yum.repos.d/*
+#%config(noreplace) %{_sysconfdir}/yum.repos.d/*
```
Comment (domcleal) [inline foreman.spec:134]: Why's this commented?

File: foreman.spec [hunk 14]
```diff
@@ -117,23 +157,13 @@ Meta Package to install requirements for ovirt support
...
-if [ $1 == 1 ];then
-cd /usr/share/foreman; rm -f Gemfile.lock; /usr/bin/bundle install --local 1>/dev/null 2>&1
-fi
-
-%postun ovirt
-if [ $1 == 0 ]; then
-cd /usr/share/foreman; rm -f Gemfile.lock; /usr/bin/bundle install --local 1>/dev/null 2>&1
-fi
-
 %package ec2
```
Comment (domcleal) [inline foreman.spec:164]: This should be renamed to match the Debian package, as per [#2123](http://theforeman.org/issues/2123).

File: foreman.spec [hunk 15]
```diff
@@ -164,23 +184,33 @@ Meta Package to install requirements for vmware support
...
-
-%postun vmware
-if [ $1 == 0 ]; then
-cd /usr/share/foreman; rm -f Gemfile.lock; /usr/bin/bundle install --local 1>/dev/null 2>&1
-fi
+%package assets
+Summary: Foreman asset pipeline support
+Group: Applications/system
+Requires: %{name} = %{version}-%{release}
+Requires: %{?scl_prefix}rubygem(coffee-rails)
```
Comment (domcleal) [inline foreman.spec]: There are a number of version requirements missing in this subpackage, which might be vital to ensure the UI functions properly.

File: foreman.spec [hunk 16]
```diff
@@ -340,32 +325,45 @@ Foreman is based on Ruby on Rails, and this package bundles Rails and all
...
+%{scl_rake} assets:precompile:all RAILS_ENV=production --trace
+rm config/database.yml
 
 %install
 rm -rf %{buildroot}
 install -d -m0755 %{buildroot}%{_datadir}/%{name}
 install -d -m0755 %{buildroot}%{_sysconfdir}/%{name}
 install -d -m0755 %{buildroot}%{_localstatedir}/lib/%{name}
+install -d -m0755 %{buildroot}%{_localstatedir}/lib/%{name}/tmp
+install -d -m0755 %{buildroot}%{_localstatedir}/lib/%{name}/tmp/pids
```
Comment (domcleal) [inline foreman.spec:364]: These two lines don't look right, what are they for?  I thought ~foreman/tmp/ was a symlink to /var/run,

File: foreman.spec [hunk 17]
```diff
@@ -340,32 +325,45 @@ Foreman is based on Ruby on Rails, and this package bundles Rails and all
...
+install -Dp -m0644 %{confdir}/logrotate %{buildroot}%{_sysconfdir}/logrotate.d/%{name}
+install -Dp -m0644 foreman.cron.d %{buildroot}%{_sysconfdir}/cron.d/%{name}
 
-install -Dp -m0644 %{confdir}/%{name}.sysconfig %{buildroot}%{_sysconfdir}/sysconfig/%{name}
-install -Dp -m0755 %{confdir}/%{name}.init %{buildroot}%{_initrddir}/%{name}
-install -Dp -m0644 %{confdir}/%{name}.logrotate %{buildroot}%{_sysconfdir}/logrotate.d/%{name}
-install -Dp -m0644 %{confdir}/%{name}.cron.d %{buildroot}%{_sysconfdir}/cron.d/%{name}
-
-install -dm 755 $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d
-install -pm 644 %{confdir}/%{name}.repo $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d
```
Comment (domcleal) [inline foreman.spec]: Where did these go?  I guess that explains why the yum.repos.d files were commented out of the -release subpackage.

File: foreman.spec [hunk 18]
```diff
@@ -395,27 +393,37 @@ ln -sv %{_localstatedir}/lib/%{name}/public %{buildroot}%{_datadir}/%{name}/publ
 # Put logs in %{_localstatedir}/log/%{name}
 ln -sv %{_localstatedir}/log/%{name} %{buildroot}%{_datadir}/%{name}/log
 
-# Put tmp files in %{_localstatedir}/run/%{name}
-ln -sv %{_localstatedir}/run/%{name} %{buildroot}%{_datadir}/%{name}/tmp
+# Put tmp files in %{_localstatedir}/lib/%{name}/tmp
+ln -sv %{_localstatedir}/lib/%{name}/tmp %{buildroot}%{homedir}/tmp
```
Comment (domcleal) [inline foreman.spec]: Why this change?  /var/lib/foreman/tmp doesn't seem appropriate for tmp files.  (related to above comment)

File: foreman.spec [hunk 19]
```diff
@@ -395,27 +393,37 @@ ln -sv %{_localstatedir}/lib/%{name}/public %{buildroot}%{_datadir}/%{name}/publ
...
 echo %{version} > %{buildroot}%{_datadir}/%{name}/VERSION
+
+# ghost files
+touch %{buildroot}%{_datadir}/%{name}/config/initializers/local_secret_token.rb
+touch %{buildroot}%{_localstatedir}/log/%{name}/cron.log
```
Comment (domcleal) [inline foreman.spec]: We weren't going to do this, as per comments on https://github.com/theforeman/foreman-rpms/pull/34.

File: foreman.spec [hunk 20]
```diff
@@ -395,27 +393,37 @@ ln -sv %{_localstatedir}/lib/%{name}/public %{buildroot}%{_datadir}/%{name}/publ
...
+%exclude %{_datadir}/%{name}/app/assets
 %{_initrddir}/%{name}
 %config(noreplace) %{_sysconfdir}/%{name}
 %config(noreplace) %{_sysconfdir}/sysconfig/%{name}
 %config(noreplace) %{_sysconfdir}/logrotate.d/%{name}
 %config %{_sysconfdir}/cron.d/%{name}
 %attr(-,%{name},%{name}) %{_localstatedir}/lib/%{name}
-%attr(-,%{name},%{name}) %{_localstatedir}/log/%{name}
+%attr(750,%{name},%{name}) %{_localstatedir}/log/%{name}
+%ghost %{_localstatedir}/log/%{name}/cron.log
```
Comment (domcleal) [inline foreman.spec]: ditto above, shouldn't be owning one arbitrary log file.

File: foreman.spec [hunk 21]
```diff
@@ -457,22 +465,18 @@ if [ ! -d $varlibdir/%{name} -a -d $datadir/tmp -a ! -L $datadir/tmp ]; then
...
-if [ ! -f %{_datadir}/%{name}/config/initializers/local_secret_token.rb ]; then
-  touch %{_datadir}/%{name}/config/initializers/local_secret_token.rb
-  chmod 0640 %{_datadir}/%{name}/config/initializers/local_secret_token.rb
-  chgrp foreman %{_datadir}/%{name}/config/initializers/local_secret_token.rb
-  rake -f %{_datadir}/%{name}/Rakefile security:generate_token >/dev/null 2>&1 || :
-fi
 /sbin/chkconfig --add %{name} || :
 (/sbin/service foreman status && /sbin/service foreman restart) >/dev/null 2>&1
+[ -f %{homedir}/config/initializers/local_secret_token.rb ] || \
+  %{scl_rake} -f %{homedir}/Rakefile security:generate_token >/dev/null 2>&1 || :
```
Comment (domcleal) [inline foreman.spec]: Can you check the permissions on this are correct?  The code removed above was making sure this was group-readable to foreman, yet owned by root.

File: foreman.spec [hunk 22]
```diff
@@ -486,23 +490,862 @@ if [ $1 -ge 1 ] ; then
 fi
 
 %changelog
```
Comment (domcleal) [inline foreman.spec:500]: Rewriting history to the Katello one seems odd for us upstream, could we keep our upstream changelog?

File: foreman.spec (hunk 1 above)
Comment (skottler) [inline foreman.spec]: Nope, this is so people can seamlessly upgrade from the pre-release RPM's to finals. Greg uses a similar strategy with the debs.

One important thing to note about this version number, though, is that it will prevent people running `1.1stable` from upgrading to the nightly/RC releases because 1.19 is viewed as older than 1.1stable from RPM's perspective. This is a change that will only break for this version because going forward we'll be using "proper" pre-release versioning.

Comment (domcleal) [inline foreman.spec]: Just testing with rpmdev-vercmp and [reading this](http://fedoraproject.org/wiki/Archive:Tools/RPM/VersionComparison), it seems that `1.19` is newer than `1.1stable`.  The trouble is the reverse, as `1.19` is newer than `1.2` or `1.2.0` (my preference for the next release version).

How about `1.1.9999`?

<pre>
$ rpmdev-vercmp 1.1stable-3.el6 1.1.9999-1.el6
1.1stable-3.el6 &lt; 1.1.9999-1.el6
$ rpmdev-vercmp 1.1.9999-1.el6 1.2.0-1.el6
1.1.9999-1.el6 &lt; 1.2.0-1.el6
</pre>

Comment (skottler) [inline foreman.spec]: That's a nice little utility!

Changing it to 1.1.9999.

File: foreman.spec (hunk 2 above)
Comment (skottler) [inline foreman.spec]: Yeah, this doesn't really matter very much because tito uses the git repo to build the source tarball, but I've changed it back the old values regardless.

File: foreman.spec (hunk 6 above)
Comment (skottler) [inline foreman.spec:59]: Fixed.

File: foreman.spec (hunk 7 above)
Comment (skottler) [inline foreman.spec:65]: Fixed.

File: foreman.spec (hunk 3 above)
Comment (skottler) [inline foreman.spec:109]: There are some odd failures when running rake tasks without it. @xsuchy should have more info.

File: foreman.spec (hunk 15 above)
Comment (skottler) [inline foreman.spec]: I've updated the requirements to match what's in the gemfile

**Comment by skottler** [comment]
Updated and rebased. Thanks for the thorough review!

**Comment by domcleal** [comment]
FYI, the two things I'm still concerned about are:
- location of tmp/ (not /var/lib/foreman/tmp) and server pid files etc
- security token permissions

thanks for all the updates!

**Comment by skottler** [comment]
I should have time on Thursday to get back to this. Thanks for the update @domcleal!

**Comment by domcleal** [comment]
Thanks @skottler!  Merged as 09869c4 with a few fixups in the init.d script, comps and crontab.

**Comment by domcleal** [comment]
Correction, 3467429 in develop and 09869c4 in 1.2-stable.
